SecurityCodeScan/Analyzers/RequestValidationAnalyzer.cs

```
@@ -95,13 +97,49 @@ private void CheckValidateInput(SyntaxNodeAnalysisContext ctx, SyntaxNodeHelper
-                {
-                    ctx.ReportDiagnostic(Diagnostic.Create(Rule, expression.GetLocation()));
+                if (containingSymbol != "System.Web.Mvc.ValidateInputAttribute")
+                    continue;
+
+                ctx.ReportDiagnostic(Diagnostic.Create(Rule, expression.GetLocation()));
+                return;
+            }
+        }
+
+        private void CheckValidateInputInheritance(SymbolAnalysisContext ctx)
+        {
+            var classSymbol = (ITypeSymbol)(ctx.Symbol is IMethodSymbol ? ctx.Symbol.ContainingSymbol : ctx.Symbol);
+            if (!classSymbol.IsDerivedFrom("System.Web.Mvc.Controller"))
+                return;
+
+            if (ctx.Symbol.HasAttribute(attr => attr.AttributeClass.ToString().Equals("System.Web.Mvc.ValidateInputAttribute")))
+                return;
+
+            for (var baseType = GetBaseSymbols(ctx.Symbol); baseType != null; baseType = GetBaseSymbols(baseType))
+            {
+                var validateInputAttr =
+                    baseType.GetAttribute(attr => attr.AttributeClass.ToString().Equals("System.Web.Mvc.ValidateInputAttribute"));
+                if (validateInputAttr == null)
                     break;
-                }
+
+                var argValue = validateInputAttr.ConstructorArguments[0].Value.ToString();
+                if (argValue.Equals("true") || argValue.Equals("True"))
+                    return;
+
+                ctx.ReportDiagnostic(Diagnostic.Create(InheritanceRule, ctx.Symbol.Locations[0]));
             }
         }
 
+        private ISymbol GetBaseSymbols(ISymbol symbols)
+        {
+            switch (symbols)
+            {
+                case ITypeSymbol typeSymbol:
+                    return typeSymbol.BaseType;
+                case IMethodSymbol methodSymbol:
+                    return methodSymbol.OverriddenMethod;
+                default:
+                    return null;
```

It is better to throw an exception if unexpected symbol was supplied.